### PR TITLE
Adding cluster-etcd-operator sub command: ensure-env

### PR DESCRIPTION
**Pull request by dusk125**
This new sub command replaces the [bash implementation found here](https://github.com/openshift/cluster-etcd-operator/blob/master/bindata/etcd/pod.yaml#L35).

Since this subcommand is configurable with command line switches, we can more easily control what validations are done, and how/if they fail. Also, the error reporting is improved with the subcommand; the bash implementation echos the entire bash script so it's very difficult to see where something went wrong (see example below).

For example, in the case where the (Downward API value for NODE_IP is empty)[https://bugzilla.redhat.com/show_bug.cgi?id=1930422#c6], we can change the invocation of ensure-env to continue and try to validate the node IP address against the network interfaces instead of failing.

### CODE REVIEW

File: pkg/cmd/ensureenv/ensureenv.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+			return
+		}
+
+		for _, iface := range interfaces {
+			var (
+				addrs []net.Addr
+			)
+
+			if addrs, err = iface.Addrs(); err != nil {
+				return
```
Comment (hexfusion) [inline pkg/cmd/ensureenv/ensureenv.go:144]: nit: handle return var explicitly vs naked.

```
return err
```

File: pkg/cmd/ensureenv/ensureenv.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		//   interfaces, to see if we find a match.
+		// If we find a match, we'll consider ourselves valid, if we can't
+		//   we'll fail.
+		if interfaces, err = net.Interfaces(); err != nil {
+			return
+		}
+
+		for _, iface := range interfaces {
+			var (
+				addrs []net.Addr
```
Comment (hexfusion) [inline pkg/cmd/ensureenv/ensureenv.go:140]: nit: style wise a single var should be initialized like
```
var name type
```
just drop ().

File: pkg/cmd/ensureenv/ensureenv.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+			)
+
+			if addrs, err = iface.Addrs(); err != nil {
+				return
+			}
+
+			for _, addr := range addrs {
+				if strings.Contains(addr.String(), fallbackEV) {
```
Comment (hexfusion) [inline pkg/cmd/ensureenv/ensureenv.go:148]: careful of contains here 

```
fmt.Println(strings.Contains("192.168.1.100", "192.168.1.1"))
```

File: pkg/cmd/ensureenv/ensureenv.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+					// We found a match, we'll assume now that the value of fallbackEV is good
+					//		and we'll continue allowing etcd to start.
+					return nil
```
Comment (hexfusion) [inline pkg/cmd/ensureenv/ensureenv.go:151]: Let's print a log that explicitly describes the situation as NODE_IP was empty but ...

File: pkg/cmd/ensureenv/ensureenv.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+				}
+			}
+		}
+
+		// If we drop out of the loop, we haven't found a match and we need to error-out
+		return fmt.Errorf("Failed to find any ip addresses in network interfaces that match the value of %v", e.fallbackIPEV)
```
Comment (hexfusion) [inline pkg/cmd/ensureenv/ensureenv.go:157]: style nits:

- error message does not start with capital letter.
- "failed to find ip address on network interfaces: %v"

File: pkg/cmd/ensureenv/ensureenv.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	return
```
Comment (hexfusion) [inline pkg/cmd/ensureenv/ensureenv.go:160]: nit be explicit on returns 
```
return nil
```

File: pkg/cmd/ensureenv/ensureenv.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+		},
+	}
+
+	ensure.AddFlags(cmd.Flags())
+	return cmd
+}
+
+func (e *ensureOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&e.allowInvalidIPEV, "allow-invalid-ip-ev-val", true, "If this is set and the value of ip-ev is invalid (not equal to environment variables given by --ip-ev-equals), continue validation and check the value of --fallback-ip-ev against the network interfaces.")
+	fs.StringVar(&e.ipEV, "ip-ev", "NODE_IP", "The environment variable that should contain the node IP address.")
```
Comment (hexfusion) [inline pkg/cmd/ensureenv/ensureenv.go:77]: let's simplify and make the default empty string then validate is checking if empty and error.

Comment (hexfusion) [reply pkg/cmd/ensureenv/ensureenv.go:77]: nit: have the flag name explicitly describe the value maybe something like

```
fs.StringVar(&e.nodeIP, "node-ip".....  "The internal IP of the node provided by downward API"
```

File: pkg/cmd/ensureenv/ensureenv.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+			}
+		},
+	}
+
+	ensure.AddFlags(cmd.Flags())
+	return cmd
+}
+
+func (e *ensureOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&e.allowInvalidIPEV, "allow-invalid-ip-ev-val", true, "If this is set and the value of ip-ev is invalid (not equal to environment variables given by --ip-ev-equals), continue validation and check the value of --fallback-ip-ev against the network interfaces.")
```
Comment (hexfusion) [inline pkg/cmd/ensureenv/ensureenv.go:76]: I like the thought but let's boil this down to simplify the UX. The flags should only include inputs necessary to perform the tasks. The logic should be straightforward and simple.

File: pkg/cmd/ensureenv/ensureenv.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+	fs.StringVar(&e.ipEV, "ip-ev", "NODE_IP", "The environment variable that should contain the node IP address.")
+	fs.StringVar(&e.fallbackIPEV, "fallback-ip-ev", "NODE_NODE_ENVVAR_NAME_IP", "If the value of the environment variable given by --ip-ev is not set (or is empty) and --allow-invalid-ip-ev-val is true, use this environment variable to check against the network interfaces to find the nodes ip address.")
```
Comment (hexfusion) [inline pkg/cmd/ensureenv/ensureenv.go:78]: same as above let the default be empty to simplify validate.

```
current-revision-node-ip
```

**Review by hexfusion — COMMENTED** [review]
Great start a few notes on style and simplifying. In general, inputs should be explicit and only from flags. Another flag could toggle toleration of empty node-ip and best-effort match on interface IP.

But for now as David described let's prove the failure and fail fast with solid logging around the problem. This can be further justification for pressure to resolve. Great work!

We will also want to update the etcd pod spec[1] to include this.

[1] https://github.com/openshift/cluster-etcd-operator/blob/release-4.10/bindata/etcd/pod.yaml#L35

**Comment by Elbehery** [comment]
@dusk125  can u please run `make verify` on the root folder ? .. the ci is failing the verify step

**Comment by tjungblu** [comment]
great addition!
/lgtm

File: pkg/cmd/ensureenv/ensureenv.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+			}
+		},
+	}
+
+	ensure.AddFlags(cmd.Flags())
+	return cmd
+}
+
+func (e *ensureOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&e.allowInvalidNodeIP, "allow-invalid-node-ip", true, "If this is set and the value of --node-ip is invalid (not equal to environment variables given by --ip-ev-equals), continue validation and check the value of --current-revision-node-ip against the network interfaces.")
```
Comment (hasbro17) [inline pkg/cmd/ensureenv/ensureenv.go:63]: > not equal to environment variables given by --ip-ev-equals

What's `--ip-ev-equals`?

File: pkg/cmd/ensureenv/ensureenv.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+				return err
+			}
+
+			for _, addr := range addrs {
+				switch ip := addr.(type) {
+				// Casting here to remove the mask before checking for equality.
+				case *net.IPNet:
+					if ip.IP.Equal(currentRevIP) {
+						// We found a match, we'll assume now that the value of fallbackEV is good
+						//		and we'll continue allowing etcd to start.
```
Comment (hasbro17) [inline pkg/cmd/ensureenv/ensureenv.go:124]: Nit:
```suggestion
						// and we'll continue allowing etcd to start.
```

Comment (hasbro17) [reply pkg/cmd/ensureenv/ensureenv.go:124]: But you can do this in a follow up if you want since the PR is mostly good to go.

**Review by hasbro17 — COMMENTED** [review]
Looks good. Just a couple of nits.

File: pkg/cmd/ensureenv/ensureenv.go (hunk 10 above)
Comment (dusk125) [reply pkg/cmd/ensureenv/ensureenv.go]: An artifact from a previous commit, I've removed this and fixed the comment.

**Pull request merged**